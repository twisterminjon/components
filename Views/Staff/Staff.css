.staff {
  height: 100%;
  background-color: var(--staffpage_bg);
  padding-top: var(--searchbar-height);
}

.staff-searchbar {
  position: absolute;
  top: var(--header-height);
}

.staff-list {
  height: calc(100% - var(--header-height));
}
