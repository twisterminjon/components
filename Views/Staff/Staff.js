import React, { useEffect, useState } from 'react'
import PropTypes from 'prop-types'

import UserList from '../../Organisms/UserList/UserList'
import SearchBar from '../../Molecules/SearchBar/SearchBar'
import DocTitle from '../../Atoms/DocTitle/DocTitle'
import SpinnerDots from '../../Atoms/SpinnerDots/SpinnerDots'

import './Staff.css'

import debug from 'debug'
const d = debug('project:Staff')

Staff.propTypes = {
  /** Array of users to display */
  users: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      displayName: PropTypes.string.isRequired,
      profileImage: PropTypes.string,
      overallStatus: PropTypes.string.isRequired,
      favorite: PropTypes.bool.isRequired,
    })
  ),
  /** Function called when the call button is clicked */
  onStartCall: PropTypes.func.isRequired,

  /** If true, will show the message button */
  canMessage: PropTypes.bool.isRequired,

  /** If true, will show the dial button */
  canCall: PropTypes.bool,

  /** Function called when the message button is clicked */
  onMessageUser: PropTypes.func,

  /** Styles applied to wrapper */
  style: PropTypes.object,

  /** Whether or not the results are loading */
  loading: PropTypes.bool.isRequired,

  /** The current page number (starting from 0, of the paginated search results) */
  pageNumber: PropTypes.number.isRequired,

  /** Called, with the page number index, after page number has changed */
  onPageNumberChange: PropTypes.func.isRequired,
}

Staff.defaultProps = {
  style: {},
  onMessageUser: () => {},
}

export default function Staff({
  users,
  onStartCall,
  canMessage,
  canCall,
  onMessageUser,
  style,
  loading,
  pageNumber,
  onPageNumberChange,
  onSearchChange,
}) {
  const handleStartCall = user => {
    d(`calling user ${user.id}, ${user.displayName}`)
    onStartCall(user)
  }

  const handleMessageUser = user => {
    onMessageUser(user)
  }

  const [searchValue, setSearchValue] = useState('')

  useEffect(() => {
    onSearchChange(searchValue)
  }, [searchValue, onSearchChange])

  return (
    <div className="staff" style={style}>
      <DocTitle title="Staff" />
      <SearchBar onChange={setSearchValue} value={searchValue} className="staff-searchbar" />
      {
        // FIXME: Add loading spinner to UserList itself to reduce code duplication
      }
      {loading && (
        <div className="namedgroupeditor-spinner-container">
          <SpinnerDots className="namedgroupeditor-spinner" />
        </div>
      )}
      <UserList
        canMessage={canMessage}
        canCall={canCall}
        userList={users}
        onCall={handleStartCall}
        onMessage={handleMessageUser}
        className="staff-list"
        loading={loading}
        pageNumber={pageNumber}
        onPageNumberChange={onPageNumberChange}
      />
    </div>
  )
}
