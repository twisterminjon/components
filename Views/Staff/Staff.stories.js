import React from 'react'
import { MockedProvider } from '@apollo/react-testing'

import { storiesOf } from '@storybook/react'
import { withKnobs, text, boolean, select } from '@storybook/addon-knobs'
import StoryRouter from 'storybook-react-router'

import Staff from './Staff'
import UsersQl from '../../../services/UsersQl'

const mocks = [
  {
    request: {
      query: UsersQl.updateUserFavorite(),
      variables: {
        id: '2',
        favorites: { connect: [{ id: '2' }] },
      },
    },
    result: {
      data: {
        updateUser: {
          id: '2',
          displayName: 'Buck',
          favorites: { id: '3', displayName: 'Wild' },
        },
      },
    },
  },
]

const stories = storiesOf('Provider/v1/Views', module)
stories
  .addDecorator(withKnobs)
  .addDecorator(StoryRouter())
  .addDecorator(story => (
    <MockedProvider mocks={mocks} addTypeName={false}>
      {story()}
    </MockedProvider>
  ))
  .add('Staff', () => {
    let users = [
      {
        id: '123',
        displayName: 'Dr. Stephen Strange',
        profileImage: 'https://www.fillmurray.com/60/60',
        overallStatus: 'available',
        favorite: false,
      },

      {
        id: '4',
        displayName: 'Nurse Jacky',
        profileImage: 'https://www.fillmurray.com/100/100',
        overallStatus: 'busy',
        favorite: true,
      },

      {
        id: '5',
        displayName: 'Dr. Douglas Powers',
        profileImage: 'https://www.fillmurray.com/160/160',
        overallStatus: 'offline',
        favorite: false,
      },

      {
        id: '1263',
        displayName: 'Una Brow',
        profileImage: 'https://www.fillmurray.com/200/200',
        overallStatus: 'available',
        favorite: true,
      },
    ]

    const filterLab = 'filter'
    const filterDef = ''
    const filterVal = text(filterLab, filterDef)

    const canMessageLab = 'canMessage'
    const canMessageDef = true
    const canMessageVal = boolean(canMessageLab, canMessageDef)

    const userCountLab = 'userCount'
    const userCountOpts = {
      0: '0',
      1: '1',
      many: 'many',
    }
    const userCountDef = 'many'
    const userCountVal = select(userCountLab, userCountOpts, userCountDef)

    let usersVal = []
    switch (userCountVal) {
      case '0':
        usersVal = []
        break

      case '1':
        usersVal = [
          {
            id: '1263',
            displayName: 'Una Brow',
            profileImage: 'https://www.fillmurray.com/200/200',
            overallStatus: 'available',
            favorite: true,
          },
        ]
        break

      default:
        usersVal = users
    }

    return <Staff users={usersVal} filter={filterVal} onStartCall={() => {}} canMessage={canMessageVal} />
  })
