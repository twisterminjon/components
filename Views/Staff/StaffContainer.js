import React, { useCallback, useState } from 'react'
import PropTypes from 'prop-types'

import UsersQl from '../../../services/UsersQl'

import ErrorPage from '../../Views/ErrorPage/ErrorPage'
import ServerIssue from '../../Views/ServerIssue/ServerIssueContainer'
import Staff from './Staff'
import SearchPlaceholder from '../../Molecules/UserCardCall/UserCardCallPlaceholder'

import { usePaginationQuery } from '@shared/hooks'

import debug from 'debug'
const d = debug('project:SearchContainer')

StaffContainer.propTypes = {
  /** Function called to start a call */
  onStartCall: PropTypes.func.isRequired,
  /** If true, will show the message button */
  canMessage: PropTypes.bool,
  /** If true, will show the dial button */
  canCall: PropTypes.bool,
}
StaffContainer.defaultProps = {
  canMessage: null,
}

const SEARCH_QUERY = UsersQl.getStaffUsers()

export default function StaffContainer({ onStartCall, canMessage, canCall, ...rest }) {
  const [searchValue, setSearchValue] = useState(null)

  const {
    apolloContext: { loading, error, data, networkStatus },
    paginationContext: { pageNumber, setPageNumber },
  } = usePaginationQuery(
    SEARCH_QUERY,
    {
      variables: {
        displayName: searchValue,
      },
    },
    {
      mergePath: 'me.paginatedCanCallUsers',
    }
  )

  const handleUserMessage = useCallback(
    user => {
      rest.history.push(`/app/messages/user/${user.id}`)
    },
    [rest.history]
  )

  const handleSearchChange = useCallback(
    searchValue => {
      if (searchValue) {
        setPageNumber(0)
      }

      setSearchValue(searchValue)
    },
    [setPageNumber]
  )

  const users =
    (data && data.me && data.me.paginatedCanCallUsers && data.me.paginatedCanCallUsers.filter(user => user.isActive)) ||
    []

  // unhandled error
  if (error) {
    if (!error.message.includes('Network error: Failed to fetch')) {
      d(`SEARCH_QUERY failed failed on unhandled error`)
      return <ErrorPage error={error} />
    }
    if (!data && error.message.includes('Network error: Failed to fetch')) {
      d(`SEARCH_QUERY failed trying to reach api server`)
      return <ServerIssue />
    }
  }

  // fetch error (api server down) AND no data
  if (!users && loading && networkStatus !== 6) {
    d(`SEARCH_QUERY loading value ${loading}, showing placeholder`)
    return <SearchPlaceholder />
  }

  return (
    <Staff
      users={users}
      onStartCall={onStartCall}
      canMessage={canMessage}
      canCall={canCall}
      onMessageUser={handleUserMessage}
      loading={loading}
      pageNumber={pageNumber}
      onPageNumberChange={setPageNumber}
      onSearchChange={handleSearchChange}
    />
  )
}
