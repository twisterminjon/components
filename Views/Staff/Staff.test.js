import React from 'react'
import { MockedProvider } from '@apollo/react-testing'

import Staff from './Staff'
import UsersQl from '../../../services/UsersQl'

const UPDATE_FAV_MUTATION = UsersQl.updateUserFavorite()

const mocks = [
  {
    request: {
      query: UPDATE_FAV_MUTATION,
      variables: {
        id: '2',
        favorites: { disconnect: [{ id: '3' }] },
      },
    },
    result: {
      data: {
        updateUser: {
          id: '2',
          displayName: 'Buck',
          favorites: [{ id: '3', displayName: 'Wild' }],
        },
      },
    },
  },
]

let users = [
  {
    id: '123',
    displayName: 'Dr. Stephen Strange',
    profileImage: 'https://www.fillmurray.com/60/60',
    overallStatus: 'available',
    favorite: false,
  },

  {
    id: '4',
    displayName: 'Nurse Jacky',
    profileImage: 'https://www.fillmurray.com/100/100',
    overallStatus: 'busy',
    favorite: true,
  },

  {
    id: '5',
    displayName: 'Dr. Douglas Powers',
    profileImage: 'https://www.fillmurray.com/160/160',
    overallStatus: 'offline',
    favorite: false,
  },

  {
    id: '1263',
    displayName: 'Una Brow',
    profileImage: 'https://www.fillmurray.com/200/200',
    overallStatus: 'available',
    favorite: true,
  },
]

describe('Staff', () => {
  const mockFun = jest.fn()

  it('matches the snapshot', () => {
    const wrapper = window.shallow(
      <MockedProvider mocks={mocks} addTypename={false}>
        <Staff
          users={users}
          filter=""
          onStartCall={mockFun}
          canMessage={true}
          loading={false}
          pageNumber={0}
          onPageNumberChange={mockFun}
        />
      </MockedProvider>
    )

    expect(wrapper.find(Staff)).toMatchSnapshot()
  })

  it('matches the snapshot with no users', () => {
    const wrapper = window.shallow(
      <MockedProvider mocks={mocks} addTypename={false}>
        <Staff
          users={[]}
          filter=""
          onStartCall={mockFun}
          canMessage={true}
          loading={false}
          pageNumber={0}
          onPageNumberChange={mockFun}
        />
      </MockedProvider>
    )

    expect(wrapper.find(Staff)).toMatchSnapshot()
  })

  it('matches the snapshot with one user', () => {
    const wrapper = window.shallow(
      <MockedProvider mocks={mocks} addTypename={false}>
        <Staff
          users={[
            {
              id: '1263',
              displayName: 'Una Brow',
              profileImage: 'https://www.fillmurray.com/200/200',
              overallStatus: 'available',
              favorite: true,
            },
          ]}
          filter=""
          onStartCall={mockFun}
          canMessage={true}
          loading={false}
          pageNumber={0}
          onPageNumberChange={mockFun}
        />
      </MockedProvider>
    )

    expect(wrapper.find(Staff)).toMatchSnapshot()
  })

  it('matches the snapshot with a filter', () => {
    const wrapper = window.shallow(
      <MockedProvider mocks={mocks} addTypename={false}>
        <Staff
          users={users}
          filter="una"
          onStartCall={mockFun}
          canMessage={true}
          loading={false}
          pageNumber={0}
          onPageNumberChange={mockFun}
        />
      </MockedProvider>
    )

    expect(wrapper.find(Staff)).toMatchSnapshot()
  })

  it('matches the snapshot with messaging off', () => {
    const wrapper = window.shallow(
      <MockedProvider mocks={mocks} addTypename={false}>
        <Staff
          users={users}
          filter=""
          onStartCall={mockFun}
          canMessage={false}
          loading={false}
          pageNumber={0}
          onPageNumberChange={mockFun}
        />
      </MockedProvider>
    )

    expect(wrapper.find(Staff)).toMatchSnapshot()
  })

  it('matches the snapshot with when loading', () => {
    const wrapper = window.shallow(
      <MockedProvider mocks={mocks} addTypename={false}>
        <Staff
          users={users}
          filter=""
          onStartCall={mockFun}
          canMessage={false}
          loading={true}
          pageNumber={0}
          onPageNumberChange={mockFun}
        />
      </MockedProvider>
    )

    expect(wrapper.find(Staff)).toMatchSnapshot()
  })
})
