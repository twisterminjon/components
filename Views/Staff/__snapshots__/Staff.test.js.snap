// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Staff matches the snapshot 1`] = `
<Staff
  canMessage={true}
  filter=""
  loading={false}
  onMessageUser={[Function]}
  onPageNumberChange={[MockFunction]}
  onStartCall={[MockFunction]}
  pageNumber={0}
  style={Object {}}
  users={
    Array [
      Object {
        "displayName": "Dr. Stephen Strange",
        "favorite": false,
        "id": "123",
        "overallStatus": "available",
        "profileImage": "https://www.fillmurray.com/60/60",
      },
      Object {
        "displayName": "Nurse Jacky",
        "favorite": true,
        "id": "4",
        "overallStatus": "busy",
        "profileImage": "https://www.fillmurray.com/100/100",
      },
      Object {
        "displayName": "Dr. Douglas Powers",
        "favorite": false,
        "id": "5",
        "overallStatus": "offline",
        "profileImage": "https://www.fillmurray.com/160/160",
      },
      Object {
        "displayName": "Una Brow",
        "favorite": true,
        "id": "1263",
        "overallStatus": "available",
        "profileImage": "https://www.fillmurray.com/200/200",
      },
    ]
  }
/>
`;

exports[`Staff matches the snapshot with a filter 1`] = `
<Staff
  canMessage={true}
  filter="una"
  loading={false}
  onMessageUser={[Function]}
  onPageNumberChange={[MockFunction]}
  onStartCall={[MockFunction]}
  pageNumber={0}
  style={Object {}}
  users={
    Array [
      Object {
        "displayName": "Dr. Stephen Strange",
        "favorite": false,
        "id": "123",
        "overallStatus": "available",
        "profileImage": "https://www.fillmurray.com/60/60",
      },
      Object {
        "displayName": "Nurse Jacky",
        "favorite": true,
        "id": "4",
        "overallStatus": "busy",
        "profileImage": "https://www.fillmurray.com/100/100",
      },
      Object {
        "displayName": "Dr. Douglas Powers",
        "favorite": false,
        "id": "5",
        "overallStatus": "offline",
        "profileImage": "https://www.fillmurray.com/160/160",
      },
      Object {
        "displayName": "Una Brow",
        "favorite": true,
        "id": "1263",
        "overallStatus": "available",
        "profileImage": "https://www.fillmurray.com/200/200",
      },
    ]
  }
/>
`;

exports[`Staff matches the snapshot with messaging off 1`] = `
<Staff
  canMessage={false}
  filter=""
  loading={false}
  onMessageUser={[Function]}
  onPageNumberChange={[MockFunction]}
  onStartCall={[MockFunction]}
  pageNumber={0}
  style={Object {}}
  users={
    Array [
      Object {
        "displayName": "Dr. Stephen Strange",
        "favorite": false,
        "id": "123",
        "overallStatus": "available",
        "profileImage": "https://www.fillmurray.com/60/60",
      },
      Object {
        "displayName": "Nurse Jacky",
        "favorite": true,
        "id": "4",
        "overallStatus": "busy",
        "profileImage": "https://www.fillmurray.com/100/100",
      },
      Object {
        "displayName": "Dr. Douglas Powers",
        "favorite": false,
        "id": "5",
        "overallStatus": "offline",
        "profileImage": "https://www.fillmurray.com/160/160",
      },
      Object {
        "displayName": "Una Brow",
        "favorite": true,
        "id": "1263",
        "overallStatus": "available",
        "profileImage": "https://www.fillmurray.com/200/200",
      },
    ]
  }
/>
`;

exports[`Staff matches the snapshot with no users 1`] = `
<Staff
  canMessage={true}
  filter=""
  loading={false}
  onMessageUser={[Function]}
  onPageNumberChange={[MockFunction]}
  onStartCall={[MockFunction]}
  pageNumber={0}
  style={Object {}}
  users={Array []}
/>
`;

exports[`Staff matches the snapshot with one user 1`] = `
<Staff
  canMessage={true}
  filter=""
  loading={false}
  onMessageUser={[Function]}
  onPageNumberChange={[MockFunction]}
  onStartCall={[MockFunction]}
  pageNumber={0}
  style={Object {}}
  users={
    Array [
      Object {
        "displayName": "Una Brow",
        "favorite": true,
        "id": "1263",
        "overallStatus": "available",
        "profileImage": "https://www.fillmurray.com/200/200",
      },
    ]
  }
/>
`;

exports[`Staff matches the snapshot with when loading 1`] = `
<Staff
  canMessage={false}
  filter=""
  loading={true}
  onMessageUser={[Function]}
  onPageNumberChange={[MockFunction]}
  onStartCall={[MockFunction]}
  pageNumber={0}
  style={Object {}}
  users={
    Array [
      Object {
        "displayName": "Dr. Stephen Strange",
        "favorite": false,
        "id": "123",
        "overallStatus": "available",
        "profileImage": "https://www.fillmurray.com/60/60",
      },
      Object {
        "displayName": "Nurse Jacky",
        "favorite": true,
        "id": "4",
        "overallStatus": "busy",
        "profileImage": "https://www.fillmurray.com/100/100",
      },
      Object {
        "displayName": "Dr. Douglas Powers",
        "favorite": false,
        "id": "5",
        "overallStatus": "offline",
        "profileImage": "https://www.fillmurray.com/160/160",
      },
      Object {
        "displayName": "Una Brow",
        "favorite": true,
        "id": "1263",
        "overallStatus": "available",
        "profileImage": "https://www.fillmurray.com/200/200",
      },
    ]
  }
/>
`;
